### PR TITLE
skills: add security layer

### PR DESCRIPTION
**Pull request by zyga**
This branch adds support for generating files for all the security subsystems that snappy is currently using today. Those are: apparmor, seccomp, udev and dbus. Since StateEngine is not yet operational, the top-level API for the whole interface is skills.Repository.SecurityFilesForSnap() that produces information about all the files (and their contents) required to make effective skill-based security work.

There are several TODOs, mostly about using real (longish) preambles to various files.

### CODE REVIEW

File: skills/repo.go [hunk 1]
```diff
@@ -407,3 +415,98 @@ func (c bySlotSnapAndName) Less(i, j int) bool {
 func LoadBuiltInTypes(repo *Repository) error {
 	return nil
 }
+
+// SecuritySnippetsForSnap collects all of the snippets of a given security
+// system that affect a given snap.  The return value is indexed by app name
```
Comment (stevenwilkin) [inline skills/repo.go]: 2 spaces after `.`

File: skills/repo_test.go [hunk 2]
```diff
@@ -740,3 +740,79 @@ func (s *RepositorySuite) TestGrantsOfReturnsCorrectData(c *C) {
...
+		SlotSecuritySnippetCallback: func(skill *Skill, securitySystem SecuritySystem) ([]byte, error) {
+			if securitySystem == testSecurity {
+				return []byte(`consumer snippet`), nil
+			}
+			return nil, ErrUnknownSecurity
+		},
+	}
+	repo := s.emptyRepo
+	err := repo.AddType(t)
+	c.Assert(err, IsNil)
```
Comment (stevenwilkin) [inline skills/repo_test.go]: Maybe `c.Assert(repo.AddType(t), IsNil)` would make this test read more concisely

File: skills/repo.go [hunk 3]
```diff
@@ -407,3 +415,98 @@ func (c bySlotSnapAndName) Less(i, j int) bool {
...
+			continue
+		}
+		for _, app := range skill.Apps {
+			snippets[app] = append(snippets[app], snippet)
+		}
+	}
+	return snippets, nil
+}
+
+// SecurityFilesForSnap computes files that constitute all of the security permissions.
```
Comment (stevenwilkin) [inline skills/repo.go]: Is "computes files" the right way thing to say here? This method returns the paths and contents of the security files?

File: skills/repo_test.go [hunk 4]
```diff
@@ -740,3 +740,79 @@ func (s *RepositorySuite) TestGrantsOfReturnsCorrectData(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(s.testRepo.GrantsOf(s.skill.Snap, s.skill.Name), HasLen, 0)
 }
+
+// Tests for Repository.SecuritySnippetsForSnap()
+
+func (s *RepositorySuite) TestSlotSnippetsForSnapSuccess(c *C) {
+	var testSecurity SecuritySystem = "security"
```
Comment (stevenwilkin) [inline skills/repo_test.go]: Could use `const` here instead of `var`

File: skills/security.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// appArmor is a security subsystem that writes apparmor profiles.
+//
+// Each apparmor profile contains a simple <header><content><footer> structure.
+// The header specified an identifier that is relevant to the kernel. The
+// identifier can be either the full path of the executable or an abstract
+// identifier not related to the executable name.
+//
+// File containing apparmor profile has to be parsed, compiled and loaded into
```
Comment (stevenwilkin) [inline skills/security.go]: _**A** file containing **an** apparmor profile_ may make more sense

File: skills/security.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+// the running kernel using apparmor_parser. After this is done the actual file
+// is irrelevant and can be removed. To improve performance certain command
+// line options to apparmor_parser can be used to cache compiled profile across
```
Comment (stevenwilkin) [inline skills/security.go]: Need a plural here? _cache compiled profile**s**_

File: skills/security.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (aa *appArmor) footerForApp(snapName, appName string) []byte {
+	return []byte("}\n")
+}
+
+// secComp is a security subsystem that writes additional seccomp rules.
+//
+// Rules use a simple line-oriented record structure.  Each line specifies a
+// system call that is allowed.  Lines starting with "deny" specify system
```
Comment (stevenwilkin) [inline skills/security.go]: 2 spaces after `.`

File: skills/security.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+// calls that are explicitly not allowed. Lines starting with '#' are treated
+// as comments are ignored.
```
Comment (stevenwilkin) [inline skills/security.go]: s/are/and are/

File: skills/security.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+//
+// NOTE: This subsystem interacts with ubuntu-core-launcher. The launcher reads
+// a single profile from a specific path, parses it and loads a seccomp profile
+// (using Berkley packet filter as low level mechanism).
```
Comment (stevenwilkin) [inline skills/security.go]: s/as/as a/

**Comment by zyga** [comment]
Thanks @stevenwilkin -- I've corrected everything that you pointed out.

**Comment by chipaca** [comment]
LGTM. Landing.

**Pull request merged**